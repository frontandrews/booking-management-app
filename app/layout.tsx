import type { Metadata } from 'next';
import { Inter } from 'next/font/google';
import './globals.css';
import { cn } from '@/utils';
import Header from '@/components/layout/header';
import { ReduxProvider } from '@/redux/provider';
import { Toaster } from '@/components/ui/toaster';

const inter = Inter({ subsets: ['latin'] });

export const metadata: Metadata = {
  title: 'Booking Management App',
};

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html className="h-full" lang="en">
      <body className={cn(inter.className, 'h-full')}>
        <ReduxProvider>
          <Toaster />
          <Header />
          {children}
        </ReduxProvider>
      </body>
    </html>
  );
}
