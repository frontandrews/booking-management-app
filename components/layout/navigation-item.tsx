import Link from 'next/link';
import { buttonVariants } from '../ui/button';

interface NavigationItemProps {
  name: string;
  href: string;
}

export const NavigationItem = ({ name, href }: NavigationItemProps) => {
  const isExternal = href.startsWith('http') || href.startsWith('mailto');
  return isExternal ? (
    <a
      key={name}
      href={href}
      className={buttonVariants({ variant: 'ghost', className: 'text-white' })}
      target="_blank"
      rel="noopener noreferrer"
    >
      {name}
    </a>
  ) : (
    <Link
      key={name}
      href={href}
      className={buttonVariants({ variant: 'ghost', className: 'text-white' })}
    >
      {name}
    </Link>
  );
};
