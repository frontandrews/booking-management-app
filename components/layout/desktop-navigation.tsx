import { NavigationItem } from './navigation-item';

const navigation = [
  { name: 'Linkedin', href: 'https://www.linkedin.com/in/andrewsgomes/' },
  { name: 'Github', href: 'https://github.com/frontandrews' },
];

export const DesktopNavigation = () => {
  return (
    <div className="hidden lg:flex lg:gap-x-12">
      {navigation.map((item) => (
        <NavigationItem key={item.name} name={item.name} href={item.href} />
      ))}
    </div>
  );
};
