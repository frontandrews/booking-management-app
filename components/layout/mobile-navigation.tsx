import { Dialog, DialogPanel } from '@headlessui/react';
import { XMarkIcon } from '@heroicons/react/24/outline';
import { NavigationItem } from './navigation-item';
import Link from 'next/link';

interface MobileNavigationProps {
  mobileMenuOpen: boolean;
  setMobileMenuOpen: (open: boolean) => void;
  isAuthenticated: boolean;
  handleLogout: () => void;
}

const navigation = [
  { name: 'Linkedin', href: 'https://www.linkedin.com/in/andrewsgomes/' },
  { name: 'Github', href: 'https://github.com/frontandrews' },
];

export const MobileNavigation = ({
  mobileMenuOpen,
  setMobileMenuOpen,
  isAuthenticated,
  handleLogout,
}: MobileNavigationProps) => {
  return (
    <Dialog
      open={mobileMenuOpen}
      onClose={setMobileMenuOpen}
      className="lg:hidden"
    >
      <div className="fixed inset-0 z-50" />
      <DialogPanel className="fixed inset-y-0 right-0 z-50 w-full overflow-y-auto bg-primary px-6 py-6 sm:max-w-sm sm:ring-1 sm:ring-white/10">
        <div className="flex items-center justify-between">
          <Link
            href={isAuthenticated ? '/bookings' : '/'}
            className="-m-1.5 p-1.5 text-2xl font-bold leading-6 text-white"
          >
            HostSoft
          </Link>
          <button
            type="button"
            onClick={() => setMobileMenuOpen(false)}
            className="-m-2.5 rounded-md p-2.5 text-gray-400"
          >
            <span className="sr-only">Close menu</span>
            <XMarkIcon aria-hidden="true" className="h-6 w-6" />
          </button>
        </div>
        <div className="mt-6 flow-root">
          <div className="-my-6 divide-y divide-gray-500/25">
            <div className="space-y-2 py-6">
              {navigation.map((item) => (
                <NavigationItem
                  key={item.name}
                  name={item.name}
                  href={item.href}
                />
              ))}
            </div>
            <div className="py-6 text-right">
              {!isAuthenticated ? (
                <Link
                  href="/sign-in"
                  className="-mx-3 block rounded-lg px-3 py-2.5 text-base font-semibold leading-7 text-white hover:bg-gray-800"
                >
                  Log in
                </Link>
              ) : (
                <div
                  onClick={handleLogout}
                  className="text-sm font-semibold leading-6 text-white cursor-pointer"
                >
                  Logout <span aria-hidden="true">&rarr;</span>
                </div>
              )}
            </div>
          </div>
        </div>
      </DialogPanel>
    </Dialog>
  );
};
