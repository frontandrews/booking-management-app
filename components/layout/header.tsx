'use client';

import { useEffect, useState } from 'react';
import { useRouter, usePathname } from 'next/navigation';
import Link from 'next/link';
import { logOut } from '@/redux/features/auth/slice';
import { RootState, AppDispatch } from '@/redux/store';
import { Bars3Icon } from '@heroicons/react/24/outline';
import { DesktopNavigation } from './desktop-navigation';
import { MobileNavigation } from './mobile-navigation';
import { useAppDispatch, useAppSelector } from '@/redux/hooks';

export default function Header() {
  const router = useRouter();
  const pathname = usePathname();
  const isAuthenticated = useAppSelector(
    (state: RootState) => state.auth.user.isAuthenticated,
  );
  const dispatch = useAppDispatch<AppDispatch>();
  const [hydrated, setHydrated] = useState(false);
  const [mobileMenuOpen, setMobileMenuOpen] = useState(false);

  useEffect(() => {
    setHydrated(true);
  }, []);

  useEffect(() => {
    if (!hydrated) return;

    if (!isAuthenticated && pathname !== '/sign-in' && pathname !== '/') {
      router.push('/sign-in');
    } else if (isAuthenticated && pathname === '/') {
      router.push('/bookings');
    }
  }, [isAuthenticated, router, pathname, hydrated]);

  const handleLogout = () => {
    dispatch(logOut());
    router.push('/sign-in');
  };

  return (
    <header className="bg-primary">
      <nav
        aria-label="Global"
        className="slide-up flex items-center justify-between p-6 lg:px-8  container mx-auto"
      >
        <div className="flex lg:flex-1">
          <Link
            href={isAuthenticated ? '/bookings' : '/'}
            className="-m-1.5 p-1.5 text-2xl font-bold leading-6 text-white"
          >
            HostSoft
          </Link>
        </div>
        <div className="flex lg:hidden">
          <button
            type="button"
            onClick={() => setMobileMenuOpen(true)}
            className="-m-2.5 inline-flex items-center justify-center rounded-md p-2.5 text-gray-400"
          >
            <span className="sr-only">Open main menu</span>
            <Bars3Icon aria-hidden="true" className="h-6 w-6" />
          </button>
        </div>
        <DesktopNavigation />
        <div className="hidden lg:flex lg:flex-1 lg:justify-end">
          {hydrated && !isAuthenticated ? (
            <Link
              href="/sign-in"
              className="text-sm font-semibold leading-6 text-white"
            >
              Log in <span aria-hidden="true">&rarr;</span>
            </Link>
          ) : (
            hydrated && (
              <div
                onClick={handleLogout}
                className="text-sm font-semibold leading-6 text-white cursor-pointer"
              >
                Logout <span aria-hidden="true">&rarr;</span>
              </div>
            )
          )}
        </div>
      </nav>
      <MobileNavigation
        mobileMenuOpen={mobileMenuOpen}
        setMobileMenuOpen={setMobileMenuOpen}
        isAuthenticated={isAuthenticated}
        handleLogout={handleLogout}
      />
    </header>
  );
}
