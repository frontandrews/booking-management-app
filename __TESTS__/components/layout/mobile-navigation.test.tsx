import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { MobileNavigation } from '@/components/layout/mobile-navigation';

jest.mock('@/components/layout/navigation-item', () => ({
  NavigationItem: ({ name, href }: { name: string; href: string }) => (
    <a href={href} data-testid={`nav-item-${name}`}>
      {name}
    </a>
  ),
}));

describe('MobileNavigation Component', () => {
  const handleLogout = jest.fn();
  const setMobileMenuOpen = jest.fn();

  const renderComponent = (mobileMenuOpen: boolean, isAuthenticated: boolean) =>
    render(
      <MobileNavigation
        mobileMenuOpen={mobileMenuOpen}
        setMobileMenuOpen={setMobileMenuOpen}
        isAuthenticated={isAuthenticated}
        handleLogout={handleLogout}
      />,
    );

  test('renders navigation items correctly', () => {
    const { container } = renderComponent(true, false);

    const linkedinItem = screen.getByTestId('nav-item-Linkedin');
    expect(linkedinItem).toBeInTheDocument();
    expect(linkedinItem).toHaveTextContent('Linkedin');
    expect(linkedinItem).toHaveAttribute(
      'href',
      'https://www.linkedin.com/in/andrewsgomes/',
    );

    const githubItem = screen.getByTestId('nav-item-Github');
    expect(githubItem).toBeInTheDocument();
    expect(githubItem).toHaveTextContent('Github');
    expect(githubItem).toHaveAttribute(
      'href',
      'https://github.com/frontandrews',
    );

    expect(container).toMatchSnapshot();
  });

  test('renders HostSoft link correctly', () => {
    const { container } = renderComponent(true, false);

    const hostSoftLink = screen.getByText('HostSoft');
    expect(hostSoftLink).toBeInTheDocument();
    expect(hostSoftLink).toHaveAttribute('href', '/');

    expect(container).toMatchSnapshot();
  });

  test('renders login link when not authenticated', () => {
    const { container } = renderComponent(true, false);

    const loginLink = screen.getByText('Log in');
    expect(loginLink).toBeInTheDocument();
    expect(loginLink).toHaveAttribute('href', '/sign-in');

    expect(container).toMatchSnapshot();
  });

  test('renders logout button when authenticated', () => {
    const { container } = renderComponent(true, true);

    const logoutButton = screen.getByText('Logout');
    expect(logoutButton).toBeInTheDocument();
    fireEvent.click(logoutButton);
    expect(handleLogout).toHaveBeenCalledTimes(1);

    expect(container).toMatchSnapshot();
  });

  test('closes menu when close button is clicked', () => {
    const { container } = renderComponent(true, false);

    const closeButton = screen.getByRole('button', { name: 'Close menu' });
    expect(closeButton).toBeInTheDocument();
    fireEvent.click(closeButton);
    expect(setMobileMenuOpen).toHaveBeenCalledWith(false);

    expect(container).toMatchSnapshot();
  });

  test('does not render navigation when menu is closed', () => {
    const { container } = renderComponent(false, false);

    expect(screen.queryByTestId('nav-item-Linkedin')).not.toBeInTheDocument();
    expect(screen.queryByTestId('nav-item-Github')).not.toBeInTheDocument();
    expect(screen.queryByText('HostSoft')).not.toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
});
