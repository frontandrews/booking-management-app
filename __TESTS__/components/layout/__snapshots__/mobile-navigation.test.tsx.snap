// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileNavigation Component closes menu when close button is clicked 1`] = `<div />`;

exports[`MobileNavigation Component does not render navigation when menu is closed 1`] = `<div />`;

exports[`MobileNavigation Component renders HostSoft link correctly 1`] = `<div />`;

exports[`MobileNavigation Component renders login link when not authenticated 1`] = `<div />`;

exports[`MobileNavigation Component renders logout button when authenticated 1`] = `<div />`;

exports[`MobileNavigation Component renders navigation items correctly 1`] = `
<div
  aria-hidden="true"
/>
`;
